src/main/java/com/nutomic/syncthingandroid/syncthing/SyncthingRunnable.java

```diff
@@ -22,39 +22,77 @@
+        generate, // Generate keys, a config file and immediately exit.
+        main,     // Run the main Syncthing application.
+        reset,    // Clear the Syncthing index.
+        manual,   // Manual command
+    }
     /**
      * Constructs instance.
      *
-     * @param command The exact command to be executed on the shell.
+     * @param command Which type of Syncthing command to execute.
      */
-    public SyncthingRunnable(Context context, String command) {
+    public SyncthingRunnable(Context context, Command command) {
         mContext = context;
         mCommand = command;
     }
 
+    /**
+     * Constructs instance.
+     *
+     * @param manualCommand The exact command to be executed on the shell. Used for tests only.
+     */
+    public SyncthingRunnable(Context context, String manualCommand) {
+        mContext = context;
+        mCommand = Command.manual;
+        mManualCommand = manualCommand;
+    }
+
     @Override
     public void run() {
         SharedPreferences pm = PreferenceManager.getDefaultSharedPreferences(mContext);
         DataOutputStream dos = null;
         int ret = 1;
+        String command = mContext.getApplicationInfo().dataDir + "/" + SyncthingService.BINARY_NAME;
         Process process = null;
         try {
-            // Loop to handle syncthing restarts (these always have an error code of 3).
+            // Loop to handle Syncthing restarts (these always have an error code of 3).
             do {
                 process = Runtime.getRuntime().exec("sh");
                 dos = new DataOutputStream(process.getOutputStream());
-                // Set home directory to data folder for syncthing to use.
+                // Set (Android) home directory to data folder for syncthing to use.
                 dos.writeBytes("HOME=" + Environment.getExternalStorageDirectory() + " ");
                 dos.writeBytes("STTRACE=" + pm.getString("sttrace", "") + " ");
                 dos.writeBytes("STNORESTART=1 ");
                 dos.writeBytes("STNOUPGRADE=1 ");
-                // Call syncthing with -home (as it would otherwise use "~/.config/syncthing/".
-                dos.writeBytes(mCommand + " -home " + mContext.getFilesDir() + "\n");
-                dos.writeBytes("exit\n");
+                switch (mCommand) {
+                    case generate:
+                        dos.writeBytes(command + " -generate='" + mContext.getFilesDir() + "' -no-browser");
+                        break;
+                    case main:
+                        dos.writeBytes(command + " -home='" + mContext.getFilesDir() + "' -no-browser");
+                        break;
+                    case reset:
+                        dos.writeBytes(command + " -home='" + mContext.getFilesDir() + "' -reset");
+                        break;
+                    case manual:
+                        dos.writeBytes(mManualCommand);
+                        break;
+                    default:
```

Comment:
I would move the switch into a static method `SyncthingRunnable(Context, Command)`, then call `this(Context, f(Command))`. Makes everything simpler.